src/sql/src/pure.rs

```diff
@@ -384,7 +385,213 @@ pub async fn purify_create_source(
+                            item: name.to_string(),
+                        };
+                        // The zero-th output is the main output
+                        // TODO(petrosagg): these plus ones are an accident waiting to happen. Find a way
+                        // to handle the main source and the subsources uniformly
+                        available_subsources.insert(name, (i + 1, desc.clone()));
+                    }
+                    Some(available_subsources)
+                }
+                mz_sql_parser::ast::LoadGenerator::Counter => None,
+            };
+
+            let mut targeted_subsources = vec![];
+
+            let mut validated_requested_subsources = vec![];
+            match requested_subsources {
+                Some(CreateSourceSubsources::All) => {
+                    let available_subsources = match available_subsources {
+                        Some(available_subsources) => available_subsources,
+                        None => bail!("FOR ALL TABLES is only valid for multi-output sources"),
+                    };
+                    for (name, (_, desc)) in available_subsources {
+                        let upstream_name = UnresolvedObjectName::from(name.clone());
+                        let subsource_name = UnresolvedObjectName::unqualified(&name.item);
+                        validated_requested_subsources.push((upstream_name, subsource_name, desc));
+                    }
+                }
+                Some(CreateSourceSubsources::Subset(selected_subsources)) => {
+                    let available_subsources = match available_subsources {
+                        Some(available_subsources) => available_subsources,
+                        None => bail!("FOR TABLES (..) is only valid for multi-output sources"),
+                    };
+                    // The user manually selected a subset of upstream tables so we need to
+                    // validate that the names actually exist and are not ambiguous
+
+                    // An index from table name -> schema name -> database name -> PostgresTableDesc
+                    let mut tables_by_name = HashMap::new();
+                    for (subsource_name, (_, desc)) in available_subsources {
+                        let database = match &subsource_name.database {
+                            RawDatabaseSpecifier::Name(database) => database.clone(),
+                            RawDatabaseSpecifier::Ambient => unreachable!(),
+                        };
+                        tables_by_name
+                            .entry(subsource_name.item.clone())
+                            .or_insert_with(HashMap::new)
+                            .entry(subsource_name.schema.clone())
+                            .or_insert_with(HashMap::new)
+                            .entry(database)
+                            .or_insert(desc);
+                    }
+
+                    for subsource in selected_subsources {
```

Comment:
I think this code is complex enough that it warrants sharing an implementation with PG sources; just to make sure it was tractable, I typed it up (naming/placement could use work). If you disagree, feel free to drop on the floor:

```diff
diff --git a/src/sql/src/pure.rs b/src/sql/src/pure.rs
index a17deae26..e618f86c5 100644
--- a/src/sql/src/pure.rs
+++ b/src/sql/src/pure.rs
@@ -54,6 +54,70 @@ use crate::names::{Aug, FullObjectName, RawDatabaseSpecifier, ResolvedObjectName
 use crate::normalize;
 use crate::plan::StatementContext;
 
+fn subsource_gen<'a, T>(
+    selected_subsources: &mut Vec<CreateSourceSubsource<Aug>>,
+    tables_by_name: HashMap<String, HashMap<String, HashMap<String, &'a T>>>,
+) -> Result<Vec<(UnresolvedObjectName, UnresolvedObjectName, &'a T)>, anyhow::Error> {
+    let mut validated_requested_subsources = vec![];
+
+    for subsource in selected_subsources {
+        let (upstream_name, subsource_name) = match subsource.clone() {
+            CreateSourceSubsource::Bare(name) => {
+                let upstream_name = normalize::unresolved_object_name(name)?;
+                let subsource_name = UnresolvedObjectName::unqualified(&upstream_name.item);
+                (upstream_name, subsource_name)
+            }
+            CreateSourceSubsource::Aliased(name, alias) => {
+                (normalize::unresolved_object_name(name)?, alias)
+            }
+            CreateSourceSubsource::Resolved(_, _) => {
+                bail!("Cannot alias subsource using `INTO`, use `AS` instead")
+            }
+        };
+
+        let schemas = match tables_by_name.get(&upstream_name.item) {
+            Some(schemas) => schemas,
+            None => bail!("table {upstream_name} not found in source"),
+        };
+
+        let schema = match &upstream_name.schema {
+            Some(schema) => schema,
+            None => match schemas.keys().exactly_one() {
+                Ok(schema) => schema,
+                Err(_) => {
+                    bail!("table {upstream_name} is ambiguous, consider specifying the schema")
+                }
+            },
+        };
+
+        let databases = match schemas.get(schema) {
+            Some(databases) => databases,
+            None => bail!("schema {schema} not found in source"),
+        };
+
+        let database = match &upstream_name.database {
+            Some(database) => database,
+            None => match databases.keys().exactly_one() {
+                Ok(database) => database,
+                Err(_) => {
+                    bail!("table {upstream_name} is ambiguous, consider specifying the database")
+                }
+            },
+        };
+
+        let desc = match databases.get(database) {
+            Some(desc) => desc.clone(),
+            None => bail!("database {database} not found source"),
+        };
+
+        let qualified_upstream_name =
+            UnresolvedObjectName::qualified(&[database, schema, &upstream_name.item]);
+        validated_requested_subsources.push((qualified_upstream_name, subsource_name, desc));
+    }
+
+    Ok(validated_requested_subsources)
+}
+
 /// Purifies a statement, removing any dependencies on external state.
 ///
 /// See the section on [purification](crate#purification) in the crate
@@ -250,65 +314,10 @@ pub async fn purify_create_source(
                             .or_insert(table);
                     }
 
-                    for subsource in subsources {
-                        let (upstream_name, subsource_name) = match subsource.clone() {
-                            CreateSourceSubsource::Bare(name) => {
-                                let upstream_name = normalize::unresolved_object_name(name)?;
-                                let subsource_name =
-                                    UnresolvedObjectName::unqualified(&upstream_name.item);
-                                (upstream_name, subsource_name)
-                            }
-                            CreateSourceSubsource::Aliased(name, alias) => {
-                                (normalize::unresolved_object_name(name)?, alias)
-                            }
-                            CreateSourceSubsource::Resolved(_, _) => {
-                                bail!("Cannot alias subsource using `INTO`, use `AS` instead")
-                            }
-                        };
-
-                        let schemas = match tables_by_name.get(&upstream_name.item) {
-                            Some(schemas) => schemas,
-                            None => bail!("table {upstream_name} not found in source"),
-                        };
-
-                        let schema = match &upstream_name.schema {
-                            Some(schema) => schema,
-                            None => match schemas.keys().exactly_one() {
-                                Ok(schema) => schema,
-                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the schema"),
-                            },
-                        };
-
-                        let databases = match schemas.get(schema) {
-                            Some(databases) => databases,
-                            None => bail!("schema {schema} not found in source"),
-                        };
-
-                        let database = match &upstream_name.database {
-                            Some(database) => database,
-                            None => match databases.keys().exactly_one() {
-                                Ok(database) => database,
-                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the database"),
-                            },
-                        };
-
-                        let table_desc = match databases.get(database) {
-                            Some(table_desc) => table_desc,
-                            None => bail!("database {database} not found source"),
-                        };
-
-                        let qualified_upstream_name = UnresolvedObjectName::qualified(&[
-                            database,
-                            schema,
-                            &upstream_name.item,
-                        ]);
-                        validated_requested_subsources.push((
-                            qualified_upstream_name,
-                            subsource_name,
-                            table_desc,
-                        ));
-                    }
+                    validated_requested_subsources
+                        .extend(subsource_gen(subsources, tables_by_name)?);
                 }
+
                 None => {}
             };
 
@@ -418,7 +427,7 @@ pub async fn purify_create_source(
             let mut validated_requested_subsources = vec![];
             match requested_subsources {
                 Some(CreateSourceSubsources::All) => {
-                    let available_subsources = match available_subsources {
+                    let available_subsources = match &available_subsources {
                         Some(available_subsources) => available_subsources,
                         None => bail!("FOR ALL TABLES is only valid for multi-output sources"),
                     };
@@ -429,7 +438,7 @@ pub async fn purify_create_source(
                     }
                 }
                 Some(CreateSourceSubsources::Subset(selected_subsources)) => {
-                    let available_subsources = match available_subsources {
+                    let available_subsources = match &available_subsources {
                         Some(available_subsources) => available_subsources,
                         None => bail!("FOR TABLES (..) is only valid for multi-output sources"),
                     };
@@ -452,64 +461,8 @@ pub async fn purify_create_source(
                             .or_insert(desc);
                     }
 
-                    for subsource in selected_subsources {
-                        let (upstream_name, subsource_name) = match subsource.clone() {
-                            CreateSourceSubsource::Bare(name) => {
-                                let upstream_name = normalize::unresolved_object_name(name)?;
-                                let subsource_name =
-                                    UnresolvedObjectName::unqualified(&upstream_name.item);
-                                (upstream_name, subsource_name)
-                            }
-                            CreateSourceSubsource::Aliased(name, alias) => {
-                                (normalize::unresolved_object_name(name)?, alias)
-                            }
-                            CreateSourceSubsource::Resolved(_, _) => {
-                                bail!("Cannot alias subsource using `INTO`, use `AS` instead")
-                            }
-                        };
-
-                        let schemas = match tables_by_name.get(&upstream_name.item) {
-                            Some(schemas) => schemas,
-                            None => bail!("table {upstream_name} not found in source"),
-                        };
-
-                        let schema = match &upstream_name.schema {
-                            Some(schema) => schema,
-                            None => match schemas.keys().exactly_one() {
-                                Ok(schema) => schema,
-                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the schema"),
-                            },
-                        };
-
-                        let databases = match schemas.get(schema) {
-                            Some(databases) => databases,
-                            None => bail!("schema {schema} not found in source"),
-                        };
-
-                        let database = match &upstream_name.database {
-                            Some(database) => database,
-                            None => match databases.keys().exactly_one() {
-                                Ok(database) => database,
-                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the database"),
-                            },
-                        };
-
-                        let desc = match databases.get(database) {
-                            Some(desc) => desc.clone(),
-                            None => bail!("database {database} not found source"),
-                        };
-
-                        let qualified_upstream_name = UnresolvedObjectName::qualified(&[
-                            database,
-                            schema,
-                            &upstream_name.item,
-                        ]);
-                        validated_requested_subsources.push((
-                            qualified_upstream_name,
-                            subsource_name,
-                            desc,
-                        ));
-                    }
+                    validated_requested_subsources
+                        .extend(subsource_gen(selected_subsources, tables_by_name)?);
```
```